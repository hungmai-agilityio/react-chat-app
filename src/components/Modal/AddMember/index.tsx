import { ChangeEvent } from 'react';

// Interfaces
import { IProfile, IUser } from '@/interfaces';

// Components
import Checkbox from '@/components/Checkbox';
import { ItemUser } from '@/components/Item';
import Search from '@/components/Search';
import Divider from '@/components/Divider';

interface AddMemberProps {
  users: IUser[];
  profiles: Record<string, IProfile>;
  onSearch: (event: ChangeEvent<HTMLInputElement>) => void;
  onChecked: (id: string) => void;
  value: string;
  isActive: boolean;
  currentUserId: string;
  checkedUsers: string[];
}

const AddMember = ({
  users,
  profiles,
  onSearch,
  onChecked,
  value,
  isActive,
  currentUserId,
  checkedUsers
}: AddMemberProps) => {
  const filteredUsers = users.filter((user) => user.id !== currentUserId);

  return (
    <div className="max-h-modal-body overflow-hidden">
      {checkedUsers.length > 0 && (
        <div className="flex items-center p-3 gap-2 overflow-hidden hover:overflow-x-auto scrollbar">
          {users
            .filter((user) => checkedUsers.includes(user.id))
            .map((user) => {
              const profile = profiles[user.id];

              return (
                <ItemUser
                  key={user.id}
                  id={user.id}
                  name={user.userName}
                  avatar={profile?.avatar || ''}
                  isActive={isActive}
                  isItemChecked
                />
              );
            })}
        </div>
      )}
      <div className="bg-tertiary border-b-xs border-solid border-primary w-full px-4 py-1">
        <Search onChange={onSearch} value={value} />
      </div>
      <div className="h-modal-body hover:overflow-y-auto scrollbar">
        {filteredUsers.map((user) => {
          const profile = profiles[user.id];

          return (
            <div key={user.id}>
              <Checkbox
                id={user.id}
                onChange={onChecked}
                isChecked={checkedUsers.includes(user.id)}
              >
                <ItemUser
                  id={user.id}
                  name={user.userName}
                  avatar={profile?.avatar || ''}
                  isActive={isActive}
                />
              </Checkbox>
              <Divider />
            </div>
          );
        })}
      </div>
    </div>
  );
};

export default AddMember;
