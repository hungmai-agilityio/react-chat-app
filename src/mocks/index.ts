export const List_item_chat = [
  {
    id: 'sade3123-45435ft-553ksd-6754534',
    avatar: 'assets/images/mock-image.jpg',
    name: 'John',
    isActive: true,
    isGroup: true,
    message: 'Hello',
    time_stamp: new Date('2024-07-01T15:12:05.791Z').toString(),
    path: '/john'
  },
  {
    id: 'sa14345ft-553ksd-67095450j-4849',
    avatar: '',
    name: 'Marketing',
    isGroup: true,
    isGroupChat: true,
    message: 'Hello team',
    time_stamp: '2024-05-24T15:58:05.791Z',
    path: '/marketing'
  },
  {
    id: '53ksd-67w4435567435ft-55ft-553k4534',
    avatar: '',
    name: 'Ban',
    isActive: false,
    isGroup: true,
    message: 'Hello world',
    time_stamp: '2020-10-01T15:37:01.791Z',
    path: '/ban'
  }
];

export const List_item_user = [
  {
    id: '53ksd-67w4435545677-11445ds4534',
    avatar: 'assets/images/mock-image.jpg',
    name: 'John',
    isActive: true,
    path: '/john'
  },
  {
    id: '53ksd-67w4435567-43567-342341-13k4534',
    avatar: '',
    name: 'Max',
    isActive: true,
    path: '/max'
  },
  {
    id: '53ksd-67whu7du435ft-5ti5ft-553ken4534',
    avatar: '',
    name: 'Ban',
    isActive: false,
    path: '/max'
  }
];

export const memberInfo = [
  {
    id: '701796f0-0253-11ef-804b-d90ac4375227',
    name: 'Alice',
    avatar: '',
    isActive: true
  },
  {
    id: '7017234f0-0253-11ef-804b-d90ac4375227',
    name: 'Bob',
    avatar: 'assets/images/mock-image.jpg',
    isActive: true
  },
  {
    id: '701796f0-0253-11ef-804b-dr544375227',
    name: 'Charlie',
    avatar: '',
    isActive: false
  }
];

export const memberInfoSolo = [
  {
    id: '701796f0-0253-11ef-804b-d90ac4375227',
    name: 'Alice',
    avatar: '',
    isActive: true
  },
  {
    id: '7017234f0-0253-11ef-804b-d90ac4375227',
    name: 'Bob',
    avatar: 'assets/images/mock-image.jpg',
    isActive: true
  }
];

export const mockUsers = [
  {
    id: '8a741820-0253-11ef-804b-d90ac4375227',
    userName: 'Mary',
    email: 'mary@abc.com',
    password: 'Test12345',
    created_at: new Date(),
    updated_at: new Date()
  },
  {
    id: 'e7d1d160-078f-11ef-9ebd-d50e40f719a7',
    userName: 'Faker',
    email: 'faker@abc.com',
    password: 'Faker123',
    created_at: new Date(),
    updated_at: new Date()
  }
];

export const mockProfiles = {
  '701796f0-0253-11ef-804b-d90ac4375227': {
    id: '701796f0-0253-11ef-804b-d90ac4375227',
    user_id: '701796f0-0253-11ef-804b-d90ac4375227',
    avatar: '',
    phone: '123-456-7890'
  },
  '8a741820-0253-11ef-804b-d90ac4375227': {
    id: '8a741820-0253-11ef-804b-d90ac4375227',
    user_id: '8a741820-0253-11ef-804b-d90ac4375227',
    avatar: '',
    phone: '987-654-3210'
  },
  'e7d1d160-078f-11ef-9ebd-d50e40f719a7': {
    id: 'e7d1d160-078f-11ef-9ebd-d50e40f719a7',
    user_id: 'e7d1d160-078f-11ef-9ebd-d50e40f719a7',
    avatar: '',
    phone: '456-789-0123'
  }
};

export const mockDropdownItem = [
  { label: 'Item 1', value: 'item1' },
  { label: 'Item 2', value: 'item2' },
  { label: 'Item 3', value: 'item3' }
];

export const mockMessage = {
  id: '43274s-102-11ef-n741820-95h5hts325677j4562',
  message: 'Hello',
  sender: '8a741820-0253-11ef-804b-d90ac4375227',
  time_stamp: '2024-04-24T15:58:05.791Z',
  roomId: '97651-07i2-11ef-n741820-95h5hts32543543e2',
  isEdit: false
};

export const mockMessageEdit = {
  id: '93274s-29011ef-9ebd-95h5hts3256701h311',
  message: 'Hi',
  sender: '7017234f0-0253-11ef-804b-d90ac4375227',
  time_stamp: '2024-04-25T15:58:05.791Z',
  roomId: '97651-07i2-11ef-n741820-95h5hts32543543e2',
  isEdit: true
}
