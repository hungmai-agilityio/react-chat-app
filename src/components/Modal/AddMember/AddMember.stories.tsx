import { Meta, StoryObj } from '@storybook/react';

// Mock data
import { mockProfiles, mockUsers } from '@/mocks';

// Components
import Modal from '@/components/Modal';
import AddMember from '@/components/Modal/AddMember';

const meta: Meta<typeof AddMember> = {
  component: AddMember,
  tags: ['autodocs'],
  argTypes: {
    currentUserId: {
      description: 'Id of the currently logged in user'
    },
    users: {
      description: 'List user'
    },
    profiles: {
      description: 'User details by user id'
    },
    onSearch: {
      description: 'Handles searching for users in lists'
    },
    onChecked: {
      description: 'Handles check users to add to the member list'
    },
    value: {
      description: 'Input value for searching?'
    },
    isActive: {
      description: 'Check if any users are connected'
    },
    checkedUsers: {
      description: 'List of checked users'
    }
  }
};

export default meta;

type Story = StoryObj<typeof AddMember>;

export const Default: Story = {
  args: {
    users: mockUsers,
    profiles: mockProfiles,
    onSearch: () => {},
    onChecked: (id) => console.log('Checked user ID:', id),
    value: '',
    isActive: true,
    currentUserId: '701796f0-0253-11ef-804b-d90ac4375227',
    checkedUsers: []
  },
  render: (args) => (
    <Modal
      isOpen={true}
      onCloseModal={() => alert('Close Modal')}
      title="Add members"
      btnPrimary="Save"
      onClick={() => alert('Save')}
    >
      <AddMember {...args} />
    </Modal>
  )
};

export const MemberChecked: Story = {
  args: {
    users: mockUsers,
    profiles: mockProfiles,
    onSearch: () => {},
    onChecked: (id) => console.log('Checked user ID:', id),
    value: '',
    isActive: true,
    currentUserId: '701796f0-0253-11ef-804b-d90ac4375227',
    checkedUsers: ['8a741820-0253-11ef-804b-d90ac4375227']
  },
  render: (args) => (
    <Modal
      isOpen={true}
      onCloseModal={() => alert('Close Modal')}
      title="Add members"
      btnPrimary="Save"
      onClick={() => alert('Save')}
    >
      <AddMember {...args} />
    </Modal>
  )
};
