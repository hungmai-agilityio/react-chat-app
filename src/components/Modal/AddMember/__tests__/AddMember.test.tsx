import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

// Mock data
import { mockProfiles, mockUsers } from '@/mocks';

// Components
import AddMember from '@/components/Modal/AddMember';

const propsDefault = {
  users: mockUsers,
  profiles: mockProfiles,
  onChecked: jest.fn(),
  value: '',
  isActive: false,
  currentUserId: '701796f0-0253-11ef-804b-d90ac4375227',
  checkedUsers: []
};

describe('AddMember', () => {
  it('renders with default data', () => {
    render(<AddMember {...propsDefault} />);

    propsDefault.users.forEach((user) => {
      expect(screen.getByText(user.userName)).toBeInTheDocument();
    });
  });

  it('handles checked when checkbox is clicked', () => {
    render(<AddMember {...propsDefault} />);

    const checkbox = screen.getByText('Mary');
    fireEvent.click(checkbox);

    expect(propsDefault.onChecked).toHaveBeenCalledWith(
      '8a741820-0253-11ef-804b-d90ac4375227'
    );
  });

  it('handles search input and checks value', () => {
    render(<AddMember {...propsDefault} />);

    const searchInput = screen.queryByPlaceholderText(
      'Search'
    ) as HTMLInputElement;
    fireEvent.change(searchInput, { target: { value: 'Faker' } });

    expect(searchInput.value).toBe('Faker');
  });

  it('render with checked members', () => {
    render(
      <AddMember
        {...propsDefault}
        checkedUsers={[
          'e7d1d160-078f-11ef-9ebd-d50e40f719a7',
          '8a741820-0253-11ef-804b-d90ac4375227'
        ]}
      />
    );

    propsDefault.users
      .filter((user) =>
        [
          'e7d1d160-078f-11ef-9ebd-d50e40f719a7',
          '8a741820-0253-11ef-804b-d90ac4375227'
        ].includes(user.id)
      )
      .forEach((user) => {
        const elements = screen.getAllByText(user.userName);
        expect(elements.length).toBeGreaterThan(0);
        elements.forEach((element) => {
          expect(element).toBeInTheDocument();
        });
      });
  });

  it('should be render match to snapshot when uncheck', () => {
    const { container } = render(<AddMember {...propsDefault} />);

    expect(container).toMatchSnapshot();
  });

  it('should be render match to snapshot when checked', () => {
    const { container } = render(
      <AddMember
        {...propsDefault}
        checkedUsers={[
          'e7d1d160-078f-11ef-9ebd-d50e40f719a7',
          '8a741820-0253-11ef-804b-d90ac4375227'
        ]}
      />
    );

    expect(container).toMatchSnapshot();
  });
});
