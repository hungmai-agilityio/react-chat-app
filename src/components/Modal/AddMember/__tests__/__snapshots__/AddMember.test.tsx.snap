// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddMember should be render match to snapshot when checked 1`] = `
<div>
  <div
    class="max-h-modal-body overflow-hidden"
  >
    <div
      class="flex items-center p-3 gap-2 overflow-hidden hover:overflow-x-auto scrollbar"
    >
      <div
        class="flex items-center justify-between"
      >
        <div
          class="text-sm px-3 w-20"
        >
          <div
            class="relative mb-2"
          >
            <div
              class="text-primary bg-gray-300 flex items-center justify-center font-semibold rounded-full my-0 mx-auto w-avatar-sm h-avatar-sm text-lg"
              data-testid="avatar"
            >
              M
            </div>
          </div>
          <p
            class="inline-block max-w-28 whitespace-normal h-7 text-xs w-full overflow-hidden text-center"
          >
            Mary
          </p>
        </div>
      </div>
      <div
        class="flex items-center justify-between"
      >
        <div
          class="text-sm px-3 w-20"
        >
          <div
            class="relative mb-2"
          >
            <div
              class="text-primary bg-gray-300 flex items-center justify-center font-semibold rounded-full my-0 mx-auto w-avatar-sm h-avatar-sm text-lg"
              data-testid="avatar"
            >
              F
            </div>
          </div>
          <p
            class="inline-block max-w-28 whitespace-normal h-7 text-xs w-full overflow-hidden text-center"
          >
            Faker
          </p>
        </div>
      </div>
    </div>
    <div
      class="bg-tertiary border-b-xs border-solid border-primary w-full px-4 py-1"
    >
      <div
        class="relative"
      >
        <input
          class="border border-gray-300 focus:border-primary outline-none w-full h-10 pl-4 pr-9"
          placeholder="Search"
          value=""
        />
      </div>
    </div>
    <div
      class="h-modal-body hover:overflow-y-auto scrollbar"
    >
      <div>
        <div
          class="pl-3 flex items-center gap-4 w-full bg-blue-50"
          id="8a741820-0253-11ef-804b-d90ac4375227"
        >
          <input
            checked=""
            class="w-5 h-5 accent-primary"
            type="checkbox"
          />
          <label
            class="w-full"
          >
            <div
              class="flex items-center justify-between"
            >
              <div
                class="text-sm px-3 w-20 flex items-center gap-3 w-full h-item"
              >
                <div
                  class="relative"
                >
                  <div
                    class="text-primary bg-gray-300 flex items-center justify-center font-semibold rounded-full my-0 mx-auto w-avatar-sm h-avatar-sm text-lg"
                    data-testid="avatar"
                  >
                    M
                  </div>
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-circle absolute text-gray-400 border-xs border-white rounded-full top-6 left-7"
                    data-icon="circle"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512z"
                      fill="currentColor"
                    />
                  </svg>
                </div>
                <div
                  class="flex flex-col truncate"
                >
                  <p>
                    Mary
                     
                  </p>
                </div>
              </div>
            </div>
          </label>
        </div>
        <div
          class="flex items-center"
        >
          <div
            class="border-b border-gray-300 w-full"
          />
          <div
            class="border-b border-gray-300 w-full"
          />
        </div>
      </div>
      <div>
        <div
          class="pl-3 flex items-center gap-4 w-full bg-blue-50"
          id="e7d1d160-078f-11ef-9ebd-d50e40f719a7"
        >
          <input
            checked=""
            class="w-5 h-5 accent-primary"
            type="checkbox"
          />
          <label
            class="w-full"
          >
            <div
              class="flex items-center justify-between"
            >
              <div
                class="text-sm px-3 w-20 flex items-center gap-3 w-full h-item"
              >
                <div
                  class="relative"
                >
                  <div
                    class="text-primary bg-gray-300 flex items-center justify-center font-semibold rounded-full my-0 mx-auto w-avatar-sm h-avatar-sm text-lg"
                    data-testid="avatar"
                  >
                    F
                  </div>
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-circle absolute text-gray-400 border-xs border-white rounded-full top-6 left-7"
                    data-icon="circle"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512z"
                      fill="currentColor"
                    />
                  </svg>
                </div>
                <div
                  class="flex flex-col truncate"
                >
                  <p>
                    Faker
                     
                  </p>
                </div>
              </div>
            </div>
          </label>
        </div>
        <div
          class="flex items-center"
        >
          <div
            class="border-b border-gray-300 w-full"
          />
          <div
            class="border-b border-gray-300 w-full"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AddMember should be render match to snapshot when uncheck 1`] = `
<div>
  <div
    class="max-h-modal-body overflow-hidden"
  >
    <div
      class="bg-tertiary border-b-xs border-solid border-primary w-full px-4 py-1"
    >
      <div
        class="relative"
      >
        <input
          class="border border-gray-300 focus:border-primary outline-none w-full h-10 pl-4 pr-9"
          placeholder="Search"
          value=""
        />
      </div>
    </div>
    <div
      class="h-modal-body hover:overflow-y-auto scrollbar"
    >
      <div>
        <div
          class="pl-3 flex items-center gap-4 w-full"
          id="8a741820-0253-11ef-804b-d90ac4375227"
        >
          <input
            class="w-5 h-5 accent-primary"
            type="checkbox"
          />
          <label
            class="w-full"
          >
            <div
              class="flex items-center justify-between"
            >
              <div
                class="text-sm px-3 w-20 flex items-center gap-3 w-full h-item"
              >
                <div
                  class="relative"
                >
                  <div
                    class="text-primary bg-gray-300 flex items-center justify-center font-semibold rounded-full my-0 mx-auto w-avatar-sm h-avatar-sm text-lg"
                    data-testid="avatar"
                  >
                    M
                  </div>
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-circle absolute text-gray-400 border-xs border-white rounded-full top-6 left-7"
                    data-icon="circle"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512z"
                      fill="currentColor"
                    />
                  </svg>
                </div>
                <div
                  class="flex flex-col truncate"
                >
                  <p>
                    Mary
                     
                  </p>
                </div>
              </div>
            </div>
          </label>
        </div>
        <div
          class="flex items-center"
        >
          <div
            class="border-b border-gray-300 w-full"
          />
          <div
            class="border-b border-gray-300 w-full"
          />
        </div>
      </div>
      <div>
        <div
          class="pl-3 flex items-center gap-4 w-full"
          id="e7d1d160-078f-11ef-9ebd-d50e40f719a7"
        >
          <input
            class="w-5 h-5 accent-primary"
            type="checkbox"
          />
          <label
            class="w-full"
          >
            <div
              class="flex items-center justify-between"
            >
              <div
                class="text-sm px-3 w-20 flex items-center gap-3 w-full h-item"
              >
                <div
                  class="relative"
                >
                  <div
                    class="text-primary bg-gray-300 flex items-center justify-center font-semibold rounded-full my-0 mx-auto w-avatar-sm h-avatar-sm text-lg"
                    data-testid="avatar"
                  >
                    F
                  </div>
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-circle absolute text-gray-400 border-xs border-white rounded-full top-6 left-7"
                    data-icon="circle"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512z"
                      fill="currentColor"
                    />
                  </svg>
                </div>
                <div
                  class="flex flex-col truncate"
                >
                  <p>
                    Faker
                     
                  </p>
                </div>
              </div>
            </div>
          </label>
        </div>
        <div
          class="flex items-center"
        >
          <div
            class="border-b border-gray-300 w-full"
          />
          <div
            class="border-b border-gray-300 w-full"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
